### PR TITLE
Add the TwoFAKey modular to the Account project.

### PR DESCRIPTION
**Pull request by jianba**

### CODE REVIEW

**Comment by jianba** [comment]
So far, I have only completed a part of TwoFA. Because I'm not particularly clear about some of the requirements.

**Comment by jianba** [comment]
When the account project starts, the fields I have added are initialized to be empty. I don't know how to debug this.

File: Account/Aiursoft.Account.csproj [hunk 1]
```diff
@@ -18,4 +18,6 @@
   <ItemGroup>
     <ProjectReference Include="..\Pylon\Aiursoft.Pylon.csproj" />
   </ItemGroup>
+  
+  <ProjectExtensions><VisualStudio><UserProperties appsettings_1development_1json__JsonSchema="" /></VisualStudio></ProjectExtensions>
```
Comment (Anduin2017) [inline Account/Aiursoft.Account.csproj]: Remove this!

File: Account/Views/Account/TwoFactorAuthentication.cshtml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    <div class="col-md-6 pl-0">
+        <div class="card mb-3">
+            <div class="card-header">
+                <i class="fa fa-fw fa-address-card"></i>
+                @Localizer["TwoFactorAuthentication Settings"]
+            </div>
+            <ul class="list-group list-group-flush">
+                <li class="list-group-item">
+                    @if (null == @Model.NewTwoFAKey)
```
Comment (Anduin2017) [inline Account/Views/Account/TwoFactorAuthentication.cshtml]: I guess you shall remove the `@`

File: Pylon/Models/AiurUserBase.cs [hunk 3]
```diff
@@ -53,5 +54,8 @@ public void Update(UserInfoViewModel model)
         public override string Email { get; set; }
         [NotMapped]
         public override bool PhoneNumberConfirmed => !string.IsNullOrEmpty(PhoneNumber);
+        public virtual string TwoFAKey { get; set; }
```
Comment (Anduin2017) [inline Pylon/Models/AiurUserBase.cs]: I guess you shall put this key to `GatewayUser`. Not AiurUserBase.

File: Pylon/Models/AiurUserBase.cs [hunk 4]
```diff
@@ -53,5 +54,8 @@ public void Update(UserInfoViewModel model)
...
+        //public virtual string AuthenticatorUri { get; set; }
+        //public virtual string TwoFACode { get; set; }
```
Comment (Anduin2017) [inline Pylon/Models/AiurUserBase.cs]: Remove obsolete properties.

File: Pylon/Services/ToAPIServer/UserService.cs [hunk 5]
```diff
@@ -205,5 +205,47 @@ public async Task<AiurCollection<AuditLog>> ViewAuditLogAsync(string accessToken
...
+        }
+
+        public async Task<AiurValue<string>> SetTwoFAKeyAsync(string openId, string accessToken)
+        {
+            var url = new AiurUrl(_serviceLocation.Gateway, "User", "SetTwoFAKey", new SetTwoFAAddressModel
+            {
+                AccessToken = accessToken,
+                OpenId = openId
+            });
+            var result = await _http.Get(url, true);
```
Comment (Anduin2017) [inline Pylon/Services/ToAPIServer/UserService.cs]: Use Post instead.

File: Pylon/Services/ToAPIServer/UserService.cs [hunk 6]
```diff
@@ -205,5 +205,47 @@ public async Task<AiurCollection<AuditLog>> ViewAuditLogAsync(string accessToken
...
+        }
+
+        public async Task<AiurValue<string>> ResetTwoFAKeyAsync(string openId, string accessToken)
+        {
+            var url = new AiurUrl(_serviceLocation.Gateway, "User", "ResetTwoFAKey", new ViewTwoFAAddressModel
+            {
+                AccessToken = accessToken,
+                OpenId = openId
+            });
+            var result = await _http.Get(url, true);
```
Comment (Anduin2017) [inline Pylon/Services/ToAPIServer/UserService.cs]: Use post instead

File: Pylon/Models/API/UserAddressModels/SetTwoFAAddressModel.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Aiursoft.Pylon.Models.API.UserAddressModels
+{
+    public class SetTwoFAAddressModel : UserOperationAddressModel
+    {
```
Comment (Anduin2017) [inline Pylon/Models/API/UserAddressModels/SetTwoFAAddressModel.cs]: Why don't you pass the key?

File: Account/Migrations/20191005094428_AddTwoFAKeyAttributeToAccountDatabase.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace Aiursoft.Account.Migrations
+{
+    public partial class AddTwoFAKeyAttributeToAccountDatabase : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
```
Comment (Anduin2017) [inline Account/Migrations/20191005094428_AddTwoFAKeyAttributeToAccountDatabase.cs]: Account user shall not maintain the key in DB.

File: Account/Controllers/AccountController.cs [hunk 9]
```diff
@@ -363,6 +363,51 @@ public async Task<IActionResult> AuditLog()
...
+            {
+                model.Recover(user);
+                model.ModelStateValid = ModelState.IsValid;
+                return View(model);
+            }
+            await _userManager.UpdateAsync(user);
+            return RedirectToAction(nameof(TwoFactorAuthentication), new { justHaveUpdated = true });
+        }
+
+        public async Task<IActionResult> SetTwoFactorAuthentication(bool justHaveUpdated)
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Use POST method instead. Get an address model.

File: Account/Controllers/AccountController.cs [hunk 10]
```diff
@@ -363,6 +363,51 @@ public async Task<IActionResult> AuditLog()
...
+        {
+            var user = await GetCurrentUserAsync();
+            var TwoFAKey = await _userService.SetTwoFAKeyAsync(user.Id, await _appsContainer.AccessToken());
+            user.TwoFAKey = TwoFAKey.Value;
+            await _userManager.UpdateAsync(user);
+            return RedirectToAction(nameof(TwoFactorAuthentication));
+        }
+
+        public async Task<IActionResult> ResetTwoFactorAuthentication(bool justHaveUpdated)
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Use POST method instead. Get an address model.

File: Gateway/Controllers/UserController.cs [hunk 11]
```diff
@@ -246,5 +253,91 @@ public async Task<IActionResult> ViewAuditLog(UserOperationAddressModel model)
...
+        {
+            var user = await _grantChecker.EnsureGranted(model.AccessToken, model.OpenId, null);
+            return Json(new AiurValue<string>(user.TwoFAKey)
+            {
+                Code = ErrorType.Success,
+                Message = "Successfully get the target user's TwoFAKey  ."
+            });
+        }
+
+        public async Task<JsonResult> SetTwoFAKey(SetTwoFAAddressModel model)
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: Use post method instead.

File: Gateway/Controllers/UserController.cs [hunk 12]
```diff
@@ -246,5 +253,91 @@ public async Task<IActionResult> ViewAuditLog(UserOperationAddressModel model)
...
+            user.TwoFAKey = FormatKey(unformattedKey);
+            await _userManager.UpdateAsync(user);
+            return Json(new AiurValue<string>(user.TwoFAKey)
+            {
+                Code = ErrorType.Success,
+                Message = "Successfully set the user's TwoFAKey!"
+            });
+        }
+
+        public async Task<JsonResult> ResetTwoFAKey(SetTwoFAAddressModel model)
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: Use post method instead.

**Pull request closed without merging**